README.md
orgbox
======
[![Build Status](http://img.shields.io/travis/yasuhito/orgbox/develop.svg?style=flat)][travis]
[![Gitter chat](http://img.shields.io/badge/GITTER-orgbox-blue.svg?style=flat)][gitter]
[![Gittip](http://img.shields.io/gittip/yasuhito.svg?style=flat)][gittip]

[travis]: http://travis-ci.org/yasuhito/orgbox
[gitter]: https://gitter.im/yasuhito/orgbox
[gittip]: https://www.gittip.com/yasuhito/

[Mailbox][mailbox]-like task scheduling in Org.

If you have any questions, click on the gitter badge to join the conversation.

[mailbox]: http://www.mailboxapp.com/


Usage
=====

Move your cursor to a task you want to schedule and type <kbd>C-c C-s</kbd>.

[![Screenshot](https://raw.github.com/yasuhito/orgbox/develop/screenshot.png)][screenshot]
[screenshot]: https://raw.github.com/yasuhito/orgbox/develop/screenshot.png


Installation
============

The recommended way to install orgbox is via MELPA.

### MELPA

If you're an Emacs 24 user or you have a recent version of
`package.el` you can install orgbox from the
[MELPA](http://melpa.milkbox.net) repository.

```
M-x package-install orgbox
```

### el-get

orgbox is also available for installation from the
[el-get](https://github.com/dimitri/el-get) package manager.

```
M-x el-get-install orgbox
```


CHANGELOG.md
# Changelog

## develop (unreleased)


## 0.5.0 (5/29/2004)

### New features

* New functions `orgbox-now`, `orgbox-schedule-now` and `orgbox-agenda-schedule-now`.
  It's bound to <kbd>SPC</kbd>


## 0.4.0 (3/25/2004)

### New features

* New function `orgbox-schedule` for org-mode buffer.
* New function `orgbox-schedule-lator-today` for org-mode buffer.
* New function `orgbox-schedule-this-or-tomorrow-evening` for org-mode
  buffer.
* New function `orgbox-schedule-tomorrow` for org-mode buffer.
* New function `orgbox-schedule-this-or-next-weekend` for org-mode
  buffer.
* New function `orgbox-schedule-next-week` for org-mode buffer.
* New function `orgbox-schedule-in-a-month` for org-mode buffer.
* New function `orgbox-schedule-someday` for org-mode buffer.

### Changes

* Replaced function `orgbox-schedule` with `orgbox-agenda-schedule`.
* Replaced function `orgbox-schedule-later-today` with
  `orgbox-agenda-schedule-later-today`.
* Replaced function `orgbox-schedule-this-or-tomorrow-evening` with
  `orgbox-agenda-schedule-this-or-tomorrow-evening`
* Replaced function `orgbox-schedule-tomorrow` with
  `orgbox-agenda-schedule-tomorrow`
* Replaced function `orgbox-schedule-this-or-next-weekend` with
  `orgbox-agenda-schedule-this-or-next-weekend`
* Replaced function `orgbox-schedule-next-week` with
  `orgbox-agenda-schedule-next-week`
* Replaced function `orgbox-schedule-in-a-month` with
  `orgbox-agenda-schedule-in-a-month`
* Replaced function `orgbox-schedule-someday` with
  `orgbox-agenda-schedule-someday`


## 0.3.0 (3/24/2004)

### New features

* New defcustom `orgbox-start-time-of-day` allows users to customize
  the start time of day.
* New defcustom `orgbox-start-day-of-week` allows users to customize
  the start day of week.
* New defcustom `orgbox-start-time-of-weekends` allows users to
  customize the start time of weekends.
* New defcustom `orgbox-start-day-of-weekends` allows users to
  customize the start day of weekends.
* New defcustom `orgbox-start-time-of-evening` allows users to
  customize the start time of evening.
* New defcustom `orgbox-later` allows users to customize 'later'.
* New defcustom `orgbox-someday` allows users to customize 'someday'.

### Changes

* Replaced function `orgbox-later-today` with
  `orgbox-schedule-later-today`.
* Replaced function `orgbox-this-or-next-weekend` with
  `orgbox-schedule-this-or-next-weekend`.
* Replaced function `orgbox-this-or-tomorrow-evening` with
  `orgbox-schedule-this-or-tomorrow-evening`.
* Replaced function `orgbox-tomorrow` with `orgbox-schedule-tomorrow`.
* Replaced function `orgbox-next-week` with
  `orgbox-schedule-next-week`.
* Replaced function `orgbox-in-a-month` with
  `orgbox-schedule-in-a-month`.
* Replaced function `orgbox-someday` with `orgbox-schedule-someday`.


## 0.2.0 (3/21/2014)

### Changes

* Rebound 'in a month' to <kbd>C-c C-s m</kbd>


## 0.1.6 (3/21/2014)

### Bugs fixed

* #1 - Fix compilation warnings.


## 0.1.5 (3/18/2014)

### Bugs fixed

* Correct `orgbox` keybinding in the README.md


## 0.1.4 (3/18/2014)

### Changes

* Rebound `orgbox` to <kbd>C-c C-s</kbd>


vagrant/provision.sh
#!/bin/sh

ppa () {
    sudo apt-get install -y python-software-properties
    sudo apt-add-repository -y "$1"
}

apt_update () {
    sudo apt-get update -qq
}

apt () {
    sudo apt-get install -yy "$@"
}

# Silence debconf
export DEBIAN_FRONTEND='noninteractive'

# Bring in the necessary PPAs
apt_update
ppa ppa:cassou/emacs
apt_update

# Install make
apt make

# Install Emacs 24.x and Emacs snapshot
apt emacs24 emacs24-el emacs24-common-non-dfsg \
    emacs-snapshot emacs-snapshot-el

# Install Cask for Emacs dependency management
CASK_VERSION=0.5.0
CASK_DIR=/opt/cask-$CASK_VERSION
CASK_ARCHIVE=https://github.com/cask/cask/archive/v$CASK_VERSION.tar.gz
if ! [ -d "$CASK_DIR" -a -x "/$CASK_DIR/bin/cask" ]; then
    sudo rm -rf "$CASK_DIR"
    wget -O - $CASK_ARCHIVE | sudo tar xz -C /opt
    # Bring Cask into $PATH
    sudo ln -fs "$CASK_DIR/bin/cask" /usr/local/bin
fi


orgbox.el
;;; orgbox.el --- Mailbox-like task scheduling Org.

;; Copyright © 2014 Yasuhito Takamiya <yasuhito@gmail.com>

;; Author: Yasuhito Takamiya <yasuhito@gmail.com>
;; URL: https://github.com/yasuhito/orgbox
;; Keywords: org
;; Version: 0.5.0
;; Package-Requires: ((org "8.0") (cl-lib "0.5"))

;; This file is not part of Org.
;; This file is not part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 3, or (at your option)
;; any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs; see the file COPYING.  If not, write to the
;; Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
;; Boston, MA 02110-1301, USA.

;;; Commentary:
;;
;; This package defines a set of handy functions to schedule your
;; tasks like Mailbox (http://www.mailboxapp.com/).  Supported
;; scheduling methods are as follows:
;;
;; - Later Today
;; - This Evening
;; - Tomorrow
;; - This Weekend
;; - Next Week
;; - In a Month
;; - Someday
;; - Pick Date
;;
;;; Code:

(require 'cl-lib)
(require 'org-agenda)

(defvar orgbox-start-time-of-day)
(defvar orgbox-start-day-of-week)
(defvar orgbox-start-time-of-weekends)
(defvar orgbox-start-day-of-weekends)
(defvar orgbox-start-time-of-evening)
(defvar orgbox-later)
(defvar orgbox-someday)

(defgroup orgbox nil
  "Mailbox-like task scheduling in org agenda."
  :group 'org)

(defcustom orgbox-start-time-of-day "8:00"
  "What time does your day start?"
  :group 'orgbox
  :type 'string)

(defcustom orgbox-start-day-of-week "mon"
  "What day does your week start?"
  :group 'orgbox
  :type '(choice (const :tag "Monday" "mon")
                 (const :tag "Tuesday" "tue")
                 (const :tag "Wednesday" "wed")
                 (const :tag "Thursday" "thu")
                 (const :tag "Friday" "fri")
                 (const :tag "Saturday" "sat")
                 (const :tag "Sunday" "sun")))

(defcustom orgbox-start-time-of-weekends "10:00"
  "What time does your weekends start?"
  :group 'orgbox
  :type 'string)

(defcustom orgbox-start-day-of-weekends "sat"
  "What day does your weekends start?"
  :group 'orgbox
  :type '(choice (const :tag "Monday" "mon")
                 (const :tag "Tuesday" "tue")
                 (const :tag "Wednesday" "wed")
                 (const :tag "Thursday" "thu")
                 (const :tag "Friday" "fri")
                 (const :tag "Saturday" "sat")
                 (const :tag "Sunday" "sun")))

(defcustom orgbox-start-time-of-evening "18:00"
  "What time does your evening start?"
  :group 'orgbox
  :type 'string)

(defcustom orgbox-later 3
  "Specify 'later' in number of hours."
  :type '(choice (const :tag "1 Hour" 1)
                 (const :tag "2 Hours" 2)
                 (const :tag "3 Hours" 3)
                 (const :tag "4 Hours" 4)
                 (const :tag "5 Hours" 5)
                 (const :tag "6 Hours" 6)
                 (const :tag "7 Hours" 7)
                 (const :tag "8 Hours" 8)
                 (const :tag "9 Hours" 9)
                 (const :tag "10 Hours" 10)
                 (const :tag "11 Hours" 11)
                 (const :tag "12 Hours" 12)
                 (const :tag "13 Hours" 13)
                 (const :tag "14 Hours" 14)
                 (const :tag "15 Hours" 15)
                 (const :tag "16 Hours" 16)
                 (const :tag "17 Hours" 17)
                 (const :tag "18 Hours" 18)
                 (const :tag "19 Hours" 19)
                 (const :tag "20 Hours" 20)
                 (const :tag "21 Hours" 21)
                 (const :tag "22 Hours" 22)
                 (const :tag "23 Hours" 23)
                 (const :tag "24 Hours" 24))
  :group 'orgbox)

(defcustom orgbox-someday "+3m"
  "Specify 'Someday' in number of months."
  :type '(choice (const :tag "1 Month" "+1m")
                 (const :tag "2 Months" "+2m")
                 (const :tag "3 Months" "+3m")
                 (const :tag "4 Months" "+4m")
                 (const :tag "5 Months" "+5m")
                 (const :tag "6 Months" "+6m")
                 (const :tag "7 Months" "+7m")
                 (const :tag "8 Months" "+8m")
                 (const :tag "9 Months" "+9m")
                 (const :tag "10 Months" "+10m")
                 (const :tag "11 Months" "+11m")
                 (const :tag "12 Months" "+12m"))
  :group 'orgbox)


;; Later today

(defun orgbox-later-today (func)
  "Call FUNC to schedule a task for later today."
  (let ((later (format-time-string
                "%Y-%m-%d %H:%M"
                (time-add (current-time)
                          (seconds-to-time (* orgbox-later 60 60))))))
    (funcall func nil later)))

(defun orgbox-schedule-later-today ()
  "Schedule a task for later today."
  (interactive)
  (orgbox-later-today 'org-schedule))

(defun orgbox-agenda-schedule-later-today ()
  "Schedule a task for later today."
  (interactive)
  (orgbox-later-today 'org-agenda-schedule))


;; This or tomorrow evening

(defun orgbox-evening-p ()
  "Is already evening?"
  (string< orgbox-start-time-of-evening
           (format-time-string "%H:%M" (current-time))))

(defun orgbox-this-or-tomorrow-evening (func)
  "Call FUNC to Schedule a task for this or tomorrow evening."
  (if (orgbox-evening-p)
      (funcall func nil (format "+1d %s" orgbox-start-time-of-evening))
    (funcall func nil orgbox-start-time-of-evening)))

(defun orgbox-schedule-this-or-tomorrow-evening ()
  "Schedule a task for this or tomorrow evening."
  (interactive)
  (orgbox-this-or-tomorrow-evening 'org-schedule))

(defun orgbox-agenda-schedule-this-or-tomorrow-evening ()
  "Schedule a task for this or tomorrow evening."
  (interactive)
  (orgbox-this-or-tomorrow-evening 'org-agenda-schedule))


;; Tomorrow

(defun orgbox-tomorrow (func)
  "Call FUNC to schedule a task for tomorrow."
  (funcall func nil (format "+1d %s" orgbox-start-time-of-day)))

(defun orgbox-schedule-tomorrow ()
  "Schedule a task for tomorrow."
  (interactive)
  (orgbox-tomorrow 'org-schedule))

(defun orgbox-agenda-schedule-tomorrow ()
  "Schedule a task for tomorrow."
  (interactive)
  (orgbox-tomorrow 'org-agenda-schedule))


;; This or next weekend

(defun orgbox-weekend-p ()
  "Today is weekend?"
  (let ((day-of-week (calendar-day-of-week
                      (calendar-gregorian-from-absolute (org-today)))))
    (member day-of-week org-agenda-weekend-days)))

(defun orgbox-this-or-next-weekend (func)
  "Call FUNC to schedule a task for this or next weekend."
  (funcall func nil (format "+%s %s"
                            orgbox-start-day-of-weekends
                            orgbox-start-time-of-weekends)))

(defun orgbox-schedule-this-or-next-weekend ()
  "Schedule a task for this or next weekend."
  (interactive)
  (orgbox-this-or-next-weekend 'org-schedule))

(defun orgbox-agenda-schedule-this-or-next-weekend ()
  "Schedule a task for this or next weekend."
  (interactive)
  (orgbox-this-or-next-weekend 'org-agenda-schedule))


;; Next week

(defun orgbox-next-week (func)
  "Call FUNC to schedule a task for next week."
  (funcall func nil (format "+%s %s"
                            orgbox-start-day-of-week
                            orgbox-start-time-of-day)))

(defun orgbox-schedule-next-week ()
  "Schedule a task for next week."
  (interactive)
  (orgbox-next-week 'org-schedule))

(defun orgbox-agenda-schedule-next-week ()
  "Schedule a task for next week."
  (interactive)
  (orgbox-next-week 'org-agenda-schedule))


;; In a month

(defun orgbox-in-a-month (func)
  "Call FUNC to schedule a task for 1 month later."
  (funcall func nil "+1m"))

(defun orgbox-schedule-in-a-month ()
  "Schedule a task for 1 month later."
  (interactive)
  (orgbox-in-a-month 'org-schedule))

(defun orgbox-agenda-schedule-in-a-month ()
  "Schedule a task for 1 month later."
  (interactive)
  (orgbox-in-a-month 'org-agenda-schedule))


;; Someday

(defun orgbox-for-someday (func)
  "Call FUNC to schedule a task for someday."
  (interactive)
  (funcall func nil orgbox-someday))

(defun orgbox-schedule-someday ()
  "Schedule a task for someday."
  (interactive)
  (orgbox-for-someday 'org-schedule))

(defun orgbox-agenda-schedule-someday ()
  "Schedule a task for someday."
  (interactive)
  (orgbox-for-someday 'org-agenda-schedule))


;; Now

(defun orgbox-now (func)
  "Call FUNC to schedule a task to do now."
  (let ((now (format-time-string "%Y-%m-%d %H:%M" (current-time))))
    (funcall func nil now)))

(defun orgbox-schedule-now ()
  "Schedule a task to do now."
  (interactive)
  (orgbox-now 'org-schedule))

(defun orgbox-agenda-schedule-now ()
  "Schedule a task to do now."
  (interactive)
  (orgbox-now 'org-agenda-schedule))


;; Schedule commands

(defun orgbox-prompt ()
  "Prompt orgbox menu."
  (message "Schedule: [l]ater today  %s [e]vening  [t]omorrow  %s [w]eekend
          [n]ext week  in a [m]onth  [s]omeday  [p]ick date  [SPC] now  [q]uit/abort"
           (if (orgbox-evening-p) "tomorrow" "this")
           (if (orgbox-weekend-p) "next" "this")))

(defun orgbox-schedule ()
  "Schedule a task interactively."
  (interactive)
  (orgbox-prompt)
  (let ((a (read-char-exclusive)))
    (cl-case a
      (?l (call-interactively 'orgbox-schedule-later-today))
      (?e (call-interactively 'orgbox-schedule-this-or-tomorrow-evening))
      (?t (call-interactively 'orgbox-schedule-tomorrow))
      (?w (call-interactively 'orgbox-schedule-this-or-next-weekend))
      (?n (call-interactively 'orgbox-schedule-next-week))
      (?m (call-interactively 'orgbox-schedule-in-a-month))
      (?s (call-interactively 'orgbox-schedule-someday))
      (?p (call-interactively 'org-schedule))
      (32 (call-interactively 'orgbox-schedule-now))
      (?q (message "Abort"))
      (otherwise (error "Invalid key")))))

(defun orgbox-agenda-schedule ()
  "Schedule a task interactively."
  (interactive)
  (orgbox-prompt)
  (let ((a (read-char-exclusive)))
    (cl-case a
      (?l (call-interactively 'orgbox-agenda-schedule-later-today))
      (?e (call-interactively 'orgbox-agenda-schedule-this-or-tomorrow-evening))
      (?t (call-interactively 'orgbox-agenda-schedule-tomorrow))
      (?w (call-interactively 'orgbox-agenda-schedule-this-or-next-weekend))
      (?n (call-interactively 'orgbox-agenda-schedule-next-week))
      (?m (call-interactively 'orgbox-agenda-schedule-in-a-month))
      (?s (call-interactively 'orgbox-agenda-schedule-someday))
      (?p (call-interactively 'org-agenda-schedule))
      (32 (call-interactively 'orgbox-agenda-schedule-now))
      (?q (message "Abort"))
      (otherwise (error "Invalid key")))))

(org-defkey org-mode-map (kbd "C-c C-s") 'orgbox-schedule)
(org-defkey org-agenda-mode-map (kbd "C-c C-s") 'orgbox-agenda-schedule)

(provide 'orgbox)

;; Local Variables:
;; indent-tabs-mode: nil
;; End:

;;; orgbox.el ends here
